cunumeric/deferred.py

```diff
@@ -534,142 +534,233 @@ def _slice_store(k: slice, store: Store, dim: int) -> tuple[slice, Store]:
+                self.runtime,
+                base=key_store,
+                dtype=self.dtype,
+            )
+            rhs.putmask(mask, set_value)
+            return False, rhs, rhs, self
+        else:
+            out_dtype = rhs.dtype
+            # in the case this operation is called for the set_item, we
+            # return Point<N> type field that is later used for
+            # indirect copy operation
+            if is_set:
+                N = rhs.ndim
+                out_dtype = rhs.runtime.get_point_type(N)
 
-                        out = out._copy_store(out_tmp)
+            # TODO : current implementation of the ND output regions
+            # requires out.ndim == rhs.ndim. This will be fixed in the
+            # future
+            out = rhs.runtime.create_unbound_thunk(out_dtype, ndim=rhs.ndim)
+            key_dims = key.ndim  # dimension of the original key
 
-                return is_set, rhs, out, self
+            task = rhs.context.create_auto_task(
+                CuNumericOpCode.ADVANCED_INDEXING
+            )
+            task.add_output(out.base)
+            task.add_input(rhs.base)
+            task.add_input(key_store)
+            task.add_scalar_arg(is_set, bool)
+            task.add_scalar_arg(key_dims, ty.int64)
+            task.add_alignment(rhs.base, key_store)
+            task.add_broadcast(
+                rhs.base, axes=tuple(range(1, len(rhs.base.shape)))
+            )
+            task.execute()
 
-        if isinstance(key, NumPyThunk):
-            key = (key,)
+            # TODO : current implementation of the ND output regions
+            # requires out.ndim == rhs.ndim.
+            # The logic below will be removed in the future
+            out_dim = rhs.ndim - key_dims + 1
+
+            if out_dim != rhs.ndim:
+                out_tmp = out.base
+
+                if out.size == 0:
+                    out_shape = tuple(out.shape[i] for i in range(0, out_dim))
+                    out = cast(
+                        DeferredArray,
+                        self.runtime.create_empty_thunk(
+                            out_shape,
+                            out_dtype,
+                            inputs=[out],
+                        ),
+                    )
+                    if not is_set:
+                        out.fill(np.array(0, dtype=out_dtype))
+                else:
+                    for dim in range(rhs.ndim - out_dim):
+                        out_tmp = out_tmp.project(rhs.ndim - dim - 1, 0)
 
-        assert isinstance(key, tuple)
-        key = self._unpack_ellipsis(key, self.ndim)
+                    out = out._copy_store(out_tmp)
+            return is_set, rhs, out, self
+
+    def _create_indexing_array(
+        self,
+        key: Any,
+        is_set: bool = False,
+        set_value: Optional[Any] = None,
+    ) -> tuple[bool, Any, Any, Any]:
+
+        is_bool_array, lhs, key = self._has_single_boolean_array(key, is_set)
```

Comment:
Can you also move this normalization:

```
            # key is a single array of indices
            if isinstance(key, NumPyThunk):
                key = (key,)
                key = self._unpack_ellipsis(key, self.ndim)
                return False, self, key
```

out of `_has_single_boolean_array`, and have this call return a separate `bool_key`, instead of overwriting the incoming `key`? So we can completely separate the boolean and non-boolean cases, for easier understanding.

Reply:
sure, thank you